proposals/0000-quick-look-impredicativity.md

```diff
@@ -0,0 +1,181 @@
+
+## Examples
+
+Several examples can be found in the [paper draft](https://www.dropbox.com/s/hxjp28ym3lptmxw/quick-look-steps.pdf?dl=0). Let us review the main example, `(\x -> x) : ids`, where `ids :: [forall a. a -> a]` from the eyes of the "approachable" description.
+
+1. We infer the type of the head of the application, `(:) :: forall a. a -> [a] -> [a]`.
+2. We expose two arguments, which leads to the type `alpha -> [alpha] -> [alpha]` where `alpha` is a fresh unification variable.
+3. Now we do *quick look* by checking the "simple" expressions against those types:
+  a. `\x -> x` is not "simple", so nothing is done.
+  b. `ids :: [forall a. a -> a]` is checked against `[alpha]`. Since `alpha` is under a type constructor different than arrow, `alpha` *must* be `forall a. a -> a` for the expression to type check.
+4. The result of quick look is thus `alpha := forall a. a -> a`. This means that the second, real type checking phase must check:
+  a. `\x -> x` against `forall a. a -> a`,
+  b. `ids` against `[forall a. a -> a]`.
+
+Compare this example to `(\x -> x) : []`. In this case quick look does not return any impredicativity information (since `[]` does not contain any). Thus the inferred type for that expression is `[beta -> beta]`, without any impredicative polymorphism.
+
+## Effect and Interactions
+
+As discussed several times throughout this proposal, its goal is to give a clear and simple specification of impredicative instantiation within GHC. Whereas "simple" is a subjective matter, the availability of a set of typing rules defines a clear specification.
+
+### `TypeApplications`
+
+We deem the interaction with `TypeApplications` as a very important one; the goal is for our specification to benefit from user-written types as much as possible. We want `(:) @(forall a. a -> a) (\x -> x) []` to work, without the need of an additional type application in `[]`. The draft paper details the interaction between those features in depth.
+
+### `RankNTypes` and lack of contravariance
+
+The main disadvantage of making "quick look impredicativity" official as described in the [paper draft](https://www.dropbox.com/s/hxjp28ym3lptmxw/quick-look-steps.pdf?dl=0) is that it collides with *contravariance of function types*. That is, it no longer holds that:
+
+`(Int -> Int) -> R is a subtype of (forall a. a -> a) -> R`
+
+The problem is of technical nature, but without this restriction we would not be able to infer impredicative instantiations of functions like `($)`. **This restriction affects not only users of `ImpredicativeTypes`, but also users of `RankNTypes`, effectively making this proposal not 100% backwards compatible.**
```

Comment:
Explain why contravariance precludes inferring impredicative instanciations of `($)`.